google/api_core/extended_operation.py

```diff
@@ -0,0 +1,175 @@
+        self._refresh_and_update(retry)
+        return self._extended_operation.done
+
+    def cancel(self):
+        if self.done():
+            return False
+
+        self._cancel()
+        self._cancelled = True
+        return True
+
+    def cancelled(self):
+        # TODO(dovs): there is not currently a good way to determine whether the
+        # operation has been cancelled.
+        # The best we can do is manually keep track of cancellation
+        # and check for doneness.
+        if not self._cancelled:
+            return False
+
+        self._refresh_and_update()
+        return self._extended_operation.done
+
+    def _refresh_and_update(self, retry=polling.DEFAULT_RETRY):
+        if not self._extended_operation.done:
+            self._extended_operation = self._refresh(retry=retry)
+            self._handle_refreshed_operation()
+
+    def _handle_refreshed_operation(self):
+        with self._completion_lock:
+            if not self._extended_operation.done:
+                return
+
+            if self.error_code and self.error_message:
+                exception = exceptions.from_http_status(
+                    status_code=self.error_code,
+                    message=self.error_message,
+                    response=self._extended_operation,
+                )
+                self.set_exception(exception)
+            elif self.error_code or self.error_message:
+                exception = exceptions.GoogleAPICallError(
+                    f"Unexpected error {self.error_code}: {self.error_message}"
+                )
+                self.set_exception(exception)
+            else:
+                # Extended operations have no payload.
+                self.set_result(None)
+
+    @classmethod
+    def make(cls, refresh, cancel, extended_operation, **kwargs):
+        # Note: it is the caller's responsibility to set up refresh and cancel
+        # with their correct request argument.
+        # The reason for this is that the services that use Extended Operations
+        # have rpcs that look something like the following:
+        # // service.proto
+        # service MyLongService {
+        #     rpc StartLongTask(StartLongTaskRequest) returns (ExtendedOperation) {
+        #         option (google.cloud.operation_service) = "CustomOperationService";
+        #     }
+        # }
+        #
+        # service CustomOperationService {
+        #     rpc Get(GetOperationRequest) returns (ExtendedOperation) {
+        #         option (google.cloud.operation_polling_method) = true;
+        #     }
+        # }
+        #
+        # Any info needed for the poll, e.g. a name, path params, etc.
+        # is held in the request, which the initial client method is in a much
+        # better position to make made because the caller made the initial request.
+        #
+        # TL;DR: the caller sets up closures for refresh and cancel that carry
+        # the properly configured requests.
```

Comment:
Could this section be turned into a method docstring? Is the idea that most folks will create an `ExtendedOperation` through this method?

Reply:
Definitely. My intention was to have users instantiate `ExtendedOperation` (or relevant subclass) through this factory so that if more work needs to be done later on the child constructors don't need to do as much work.